eta/core/data.py

```diff
@@ -252,6 +252,42 @@ def validate_type(self, attr):
                 "Expected attribute '%s' to have type '%s'; found '%s'" %
                 (attr.name, self.type, etau.get_class_name(attr)))
 
+    def validate_attribute(self, attr):
+        '''Validates that the attribute is compliant with the schema.
+
+        Args:
+            attr: an Attribute
+
+        Raises:
+            AttributeSchemaError: if the attribute is not compliant with the
+                schema
+        '''
+        if attr.name != self.name:
+            raise AttributeSchemaError(
+                "Expected name '%s'; found '%s'" % (self.name, attr.name))
+
+        self.validate_type(attr)
+
+        if not self.is_valid_value(attr.value):
+            raise AttributeSchemaError(
+                "Value '%s' of attribute '%s' is not allowed by the "
+                "schema " % (attr.value, attr.name))
+
+    def is_valid_attribute(self, attr):
```

Comment:
@MikeJeffers FYI- there was a subtle bug/loophole in your implementation of Attribute validation: it didn't verify that the `type` of Attribute was correct when filtering.

So one could have tweaked an `AttributeSchema` by changing an attribute of given name from, say, `CategoricalAttribute` to `BooleanAttribute`, and it wouldn't have properly filtered the values according to the new type.

I fixed it.